enclave/core/sgx/td_basic.c

```diff
@@ -152,5 +152,14 @@ void td_clear(oe_sgx_td_t* td)
     /* Clear the magic number */
     td->magic = 0;
 
+    /* Clear states on td if the the state is not RUNNING, which could happen
+     * if the exception handler does not return. Otherwise, the state machine
+     * check will fail if the tcs is reused by another ecall.
+     * In other cases, it is expected an enclave application that implements
+     * thread support to clear the td states when necessary. This is because
+     * we cannot know the mapping between an ecall and a host thread. */
+    if (td->state != OE_TD_STATE_RUNNING)
```

Comment:
A TCS is bound to a host thread till the ecall exists. During exit, the td_state is cleared
https://github.com/openenclave/openenclave/blob/99b21cbf5b9cd43e3c2275443eb5afd43df3cb37/enclave/core/sgx/exit.S#L131

Therefore, do we need this extra clear?

Reply:
This is for the corner case I mentioned. This state transition could by bypassed if state is `SECOND_LEVEL_EXCEPTION_HANDLING` https://github.com/openenclave/openenclave/blob/99b21cbf5b9cd43e3c2275443eb5afd43df3cb37/enclave/core/sgx/exit.S#L124

This check was designed for making OCALLs during the exception handling. In the normal flow where the exception handler returns, the state will transited from `SECOND_LEVEL_EXCEPTION_HANDLING` to `RUNNING`.

In the corner case where the the enclave directly exits during the exception handler, the state will remain `SECOND_LEVEL_EXCEPTION_HANDLING`.

Reply:
Or do you think we should just make the td states persisted only during a single ECALL?

Reply:
Updated the PR such that the td states are only persisted during the ECALL.